processor/trackingplan.go

```diff
@@ -0,0 +1,250 @@
+package processor
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/processor/integrations"
+	"github.com/rudderlabs/rudder-server/processor/transformer"
+	"github.com/rudderlabs/rudder-server/utils/misc"
+	"github.com/rudderlabs/rudder-server/utils/types"
+	uuid "github.com/satori/go.uuid"
+	"time"
+)
+
+// extracts the events from Validation response and make them transformer ready
+func (proc *HandleT) getTransformerEvents(response transformer.ResponseT, eventsByMessageID map[string]types.SingularEventWithReceivedAt) ([]transformer.TransformerEventT, []*jobsdb.JobT) {
+	var eventsToTransform []transformer.TransformerEventT
+	var eventsToDrop []transformer.TransformerResponseT
+	var failedEventsToStore []*jobsdb.JobT
+
+	for _, validatedEvent := range response.Events {
+		updatedEvent := transformer.TransformerEventT{
+			Message:  validatedEvent.Output, // singular event
+			Metadata: validatedEvent.Metadata,
+		}
+		// report all violations irrespective of SchemaConfig
+		reportViolations(&validatedEvent)
+
+		sourceTpConfig := validatedEvent.Metadata.SourceTpConfig
+		if len(validatedEvent.ValidationErrors) == 0 || len(sourceTpConfig) == 0 {
+			eventsToTransform = append(eventsToTransform, updatedEvent)
+			continue
+		}
+
+		//set of all violationTypes in event
+		violationsByType := make(map[string]struct{})
+		for _, violation := range validatedEvent.ValidationErrors {
+			violationsByType[violation.Type] = struct{}{}
+		}
+
+		//not grouping by eventTypes as sourceConfig could differ
+		eventType, ok := validatedEvent.Output["type"].(string)
+		if !ok {
+			pkgLogger.Error("singular event type is unknown")
+		}
+
+		eventSpecificConfig := fetchEventConfig(sourceTpConfig, eventType)
+		globalConfig := fetchEventConfig(sourceTpConfig, "global")
```

Comment:
this may not be repeated by using mergedConfig in configT to compute it only once every time.

Reply:
Done.